### PR TITLE
Allow different filters per DataStream in a DataStreamAlias

### LINKED ISSUE #92050: In the Alias of data streams filter on only one data_stream ends up applying to all others
### Elasticsearch Version

8.5

### Installed Plugins

_No response_

### Java Version

_bundled_

### OS Version

any

### Problem Description

The use of alias of data streams behaves differently from the alias of indices.
When trying to apply filter to only one data_stream in the Alias, the result is that It ends up applying to both.

### Steps to Reproduce

```json
POST _aliases
{
    "actions": [
        {
            "add": {
                "index": "data_stream_1",
                "alias": "alias_1",
                "filter": {
                    "bool": {
                        "filter": [
                            {
                                "term": {
                                    "reaction": "report"
                                }
                            }
                        ]
                    }
                }
            }
        },
        {
            "add": {
                "index": "data_stream_2",
                "alias": "alias_1"
            }
        }
    ]
}
```
Per the above request, I’m trying to apply filter to only one data stream in the Alias but the result is that It ends up applying to both.

### result

```json
GET _alias/alias_1
{
  "data_stream_2": {
    "aliases": {
      "alias_1": {
        "filter": {
          "bool": {
            "filter": [
              {
                "term": {
                  "reaction": "report"
                }
              }
            ]
          }
        }
      }
    }
  },
  "data_stream_1": {
    "aliases": {
      "alias_1": {
        "filter": {
          "bool": {
            "filter": [
              {
                "term": {
                  "reaction": "report"
                }
              }
            ]
          }
        }
      }
    }
  }
}
```

### Logs (if relevant)

_No response_

### PR DESCRIPTION
**Pull request by masseyke**
Index aliases allow each index in the alias to have a different filter. Data stream aliases appear to do this, but in reality they do not. That is, if you ask to add two data streams to a data stream alias, each with a different filter, the API allows it but only keeps one of the filters. This PR makes it so that the DataStreamAlias keeps a map of DataStreams to filters so that the different filters are respected.
Closes https://github.com/elastic/elasticsearch/issues/92050

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java [hunk 1]
```diff
@@ -25,29 +28,62 @@
...
+             * 8.7.0. In this case the new dataStreamsToFilters map will be null. So we write a new dataStreamsToFilters using the existing
+             * filter value for all DataStreams in order to carry forward the previously-existing behavior.
+             */
+            Map<String, CompressedXContent> dataStreamsToFilters = (Map<String, CompressedXContent>) args[3];
+            CompressedXContent oldFilter = (CompressedXContent) args[2];
+            List<String> dataStreamNames = (List<String>) args[0];
+            if (dataStreamsToFilters == null && oldFilter != null && dataStreamNames != null) {
+                logger.info(
+                    "Reading in DataStreamAlias {} with a pre-8.7.0-style DataStream filter and using it for all DataStreams in "
+                        + "the DataStreamAlias",
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java:68]: Super minor nit

```suggestion
                    "Reading in data stream alias [{}] with a pre-8.7.0-style data stream filter and using it for all data streams in "
                        + "the data stream alias",
```

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java [hunk 2]
```diff
@@ -313,11 +398,24 @@ public void writeTo(StreamOutput out) throws IOException {
...
+                /*
+                 * Versions before 8.7 incorrectly only allowed a single filter for all datastreams, and randomly dropped all others. We
+                 * replicate that buggy behavior here if we have to write to an older node because there is no way to send multipole
+                 * filters to an older node.
+                 */
+                logger.info(
+                    "Sending DataStreamAlias {} to a node with version earlier than 8.7.0. Versions before 8.7.0 do not support "
+                        + "separate filters for each DataStream.",
+                    name
+                );
```
Comment (dakrone) [inline server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java:416]: I don't think we should log for binary serialization. I think the logging we have for XContent parsing is useful (and only executed once), but mixed version nodes could cause this to be spammed, so we should remove this.

**Review by dakrone — APPROVED** [review]
This generally looks good to me, thanks for tackling this Keith. I left a couple of very minor nits.

File: server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java (hunk 2 above)
Comment (masseyke) [reply server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamAlias.java:416]: My fear was that people would be accidentally corrupting their data stream aliases and that spamming would be better than nothing here. Is this not going to be a problem in practice?

**Pull request merged**